### PR TITLE
feat: TryFrom ptr-size

### PR DESCRIPTION
**Pull request by Alxandr**

### CODE REVIEW

File: src/traits.rs [hunk 1]
```diff
@@ -200,6 +200,49 @@ macro_rules! impl_from {
...
+            type Error = <$size as TryFrom<$prim>>::Error;
+
+            #[inline]
+            fn try_from(value: $name) -> Result<Self, Self::Error> {
+                <$size>::try_from(value.to_native())
+            }
+        }
+    };
+
+    (Into<$size:ty> for $name:ident via $prim:ty) => {
```
Comment (djkoloski) [inline src/traits.rs]: Since `$prim` isn't used here, we can shorten this to `($size:ty for $name:ident)`. I'd also prefer that this be separated into a `macro_rules! impl_into` macro.

File: src/traits.rs [hunk 2]
```diff
@@ -200,6 +200,49 @@ macro_rules! impl_from {
     };
 }
 
+macro_rules! impl_try_from_ptr_size {
+    (for $name:ident: TryFrom<$size:ty> via $prim:ty) => {
```
Comment (djkoloski) [inline src/traits.rs]: I would prefer this as `macro_rules! impl_try_from` with the arm being `($size:ty for $name:ident: $prim:ty)`. It's a little more consistent with how the other macros are written.

File: src/traits.rs [hunk 3]
```diff
@@ -200,6 +200,49 @@ macro_rules! impl_from {
...
+    };
+}
+
+macro_rules! impl_try_into_ptr_size {
+    (TryInto<isize> for $name:ident via i16) => {
+        impl_try_into_ptr_size!(Into<isize> for $name via i16);
+    };
+
+    (TryInto<usize> for $name:ident via u16) => {
+        impl_try_into_ptr_size!(Into<usize> for $name via i16);
```
Comment (djkoloski) [inline src/traits.rs]:
```suggestion
        impl_try_into_ptr_size!(Into<usize> for $name via u16);
```

File: src/lib.rs [hunk 4]
```diff
@@ -127,6 +127,14 @@ define_signed_integers! {
     i64_le i64_be: 8 i64,
     i128_le i128_be: 16 i128,
 }
+impl_try_into_ptr_size!(TryInto<isize> for i16_le via i16);
```
Comment (djkoloski) [inline src/lib.rs]: I think these can be merged into the `impl_signed_integer_traits` macro instead of being defined separately.

File: src/lib.rs [hunk 5]
```diff
@@ -152,6 +160,15 @@ define_unsigned_integers! {
     u128_le u128_be: 16 u128,
 }
 
+impl_try_into_ptr_size!(TryInto<usize> for u16_le via u16);
```
Comment (djkoloski) [inline src/lib.rs]: Same here for the unsigned integer impl macro.

**Review by djkoloski — CHANGES_REQUESTED** [review]

File: src/traits.rs (hunk 2 above)
Comment (Alxandr) [reply src/traits.rs]: This is not legal. You're not allowed to put `for` after a `ty` binding.

File: src/traits.rs (hunk 3 above)
Comment (Alxandr) [reply src/traits.rs]: Ah - stupid oversight. But as you suggested previously, I've removed the `via` clause for `Into`, so this part is gone.

File: src/lib.rs (hunk 4 above)
Comment (Alxandr) [reply src/lib.rs]: Unfortunately, as far as I could figure out, that's not the case (it's what I tried first). The issue is that once you bind a type as `$whatever:ty`, you can no longer match it to (for instance) `isize`. So there is no way to deal with the two that should be `Into` and not `TryInto`.

File: src/lib.rs (hunk 5 above)
Comment (Alxandr) [reply src/lib.rs]: Same as above, I don't believe it's possible. At least I don't know how.

File: src/traits.rs (hunk 1 above)
Comment (Alxandr) [reply src/traits.rs]: Done.

**Review by djkoloski — APPROVED** [review]
